scheduler/replica_scheduler.go

```diff
@@ -224,33 +223,31 @@ func (rcs *ReplicaScheduler) getDiskCandidates(nodeInfo map[string]*longhorn.Nod
 		}
 		if _, ok := usedNodes[nodeName]; !ok {
 			unusedNodes[nodeName] = node
-			if _, ok := usedZones[node.Status.Zone]; !ok {
-				unusedNodesInNewZones[nodeName] = node
-			}
 		}
 		if _, ok := usedZones[node.Status.Zone]; !ok {
-			nodesInUnusedZones[nodeName] = node
+			unusedNodesInUnusedZones[nodeName] = node
 		}
 	}
 
+	// In all cases, we should try to use a disk on an unused node in an unused zone first. Don't bother considering
+	// zoneSoftAntiAffinity and nodeSoftAntiAffinity settings if such disks are available.
+	diskCandidates, errors := getDiskCandidatesFromNodes(unusedNodesInUnusedZones)
+	if len(diskCandidates) > 0 {
+		return diskCandidates, nil
+	}
+	multiError.Append(errors)
+
 	switch {
 	case !zoneSoftAntiAffinity && !nodeSoftAntiAffinity:
-		diskCandidates, errors := getDiskCandidatesFromNodes(unusedNodesInNewZones)
-		if len(diskCandidates) > 0 {
-			return diskCandidates, nil
-		}
-		multiError.Append(errors)
+		fallthrough
+	// Same as the above. If we cannot schedule two replicas in the same zone, we cannot schedule them on the same node.
```

Comment:
If we cannot find a candidate from `unusedNodesInUnusedZones` above, should not this be `break`?

Reply:
> I'm assuming this is because the replica will soon no longer exist on the node, and the newly scheduled replica is a replacement. I am not aware of any reason to change this behavior. 

This is related to this ticket https://github.com/longhorn/longhorn/issues/3076. The expectation was there might be a different disk, disk-b, on the node-1 when user is trying to evict replicas from disk-a on node-1. It is not because that the replica will soon no longer exist on the node. Understand this usecase, maybe we want to rename some of the concepts `unusedNodesAfterEviction` and  `unusedNodesInUnusedZonesAfterEviction`

Reply:
I see. Thanks for the clarification @PhanLe1010! However, I'm not sure there's a better name for these concepts, despite my initial confusion. I thought about something like `unusedNodesConsideringEviction`, etc., but that just seems wordier.

Reply:
Verified https://github.com/longhorn/longhorn/issues/3076#issuecomment-961507326 after changes per @PhanLe1010's request.